File: NeetCode/heap/k-closest-points.java
class Solution {
	public int[][] kClosest(int[][] points, int k) {
		PriorityQueue<int[]> maxHeap = new PriorityQueue<>(new Comparator<int[]>(){
			@Override
			public int compare(int[] a, int[] b) {
				return dist(b) - dist(a);
			}
		});
		for (int[] p : points) {
			if (maxHeap.size() < k) {
				maxHeap.offer(p);
			} else {
				int[] top = maxHeap.peek();
				if (dist(p) < dist(top)) {
					maxHeap.poll();
					maxHeap.offer(p);
				}
			}
		}
		int[][] ans = new int[k][2];
		int i = 0;
		while (!maxHeap.isEmpty()) {
			int[] curr = maxHeap.poll();
			ans[i][0] = curr[0];
			ans[i][1] = curr[1];
			i++;
		}
		return ans;
	}

	private int dist(int[] p) {
		return p[0] * p[0] + p[1] * p[1];
	}
}

File: NeetCode/heap/k-closest-points.py
import heapq

class Solution:
    def kClosest(self, points: list[list[int]], k: int) -> list[list[int]]:
        maxHeap = []
        for x, y in points:
            dist = -(x ** 2 + y ** 2)
            heapq.heappush(maxHeap, [dist, x, y])
            if len(maxHeap) > k:
                heapq.heappop(maxHeap)

        ans = []
        while maxHeap:
            dist, x, y = heapq.heappop(maxHeap)
            ans.append([x, y])

        return ans
